File: src/main/java/fr/groupbees/midgard/transforms/BaseElementFn.java
package fr.groupbees.midgard.transforms;

import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.values.TypeDescriptor;

/**
 * Base class for the custom DoFn allowing the interaction with lifecycle methods.
 * <p>
 * This class gives shared elements like type descriptors, input and output
 *
 * <p>
 * The class that extend {@link fr.groupbees.midgard.transforms.BaseElementFn} must give input and output type descriptors and
 * uses the tuple tags to handle good output and failure.
 * </p>
 *
 * @param <InputT>  input of {@link org.apache.beam.sdk.transforms.DoFn}
 * @param <OutputT> output of {@link org.apache.beam.sdk.transforms.DoFn}
 * @author mazlum
 */
public abstract class BaseElementFn<InputT, OutputT> extends DoFn<InputT, OutputT> {

    protected transient TypeDescriptor<InputT> inputType;
    protected transient TypeDescriptor<OutputT> outputType;

    /**
     * <p>Constructor for BaseElementFn.</p>
     */
    protected BaseElementFn() {
        inputType = super.getInputTypeDescriptor();
        outputType = super.getOutputTypeDescriptor();
    }

    /**
     * <p>Constructor for BaseElementFn.</p>
     *
     * @param inputType  a {@link org.apache.beam.sdk.values.TypeDescriptor} object
     * @param outputType a {@link org.apache.beam.sdk.values.TypeDescriptor} object
     */
    protected BaseElementFn(TypeDescriptor<InputT> inputType, TypeDescriptor<OutputT> outputType) {
        this.inputType = inputType;
        this.outputType = outputType;
    }

    /**
     * <p>Getter for the field <code>inputType</code>.</p>
     *
     * @return a {@link org.apache.beam.sdk.values.TypeDescriptor} object
     */
    public TypeDescriptor<InputT> getInputTypeDescriptor() {
        return inputType;
    }

    /**
     * <p>Getter for the field <code>outputType</code>.</p>
     *
     * @return a {@link org.apache.beam.sdk.values.TypeDescriptor} object
     */
    public TypeDescriptor<OutputT> getOutputTypeDescriptor() {
        return outputType;
    }
}

File: src/main/kotlin/fr/groupbees/midgard/transforms/BaseElementFn.kt
package fr.groupbees.midgard.transforms

import org.apache.beam.sdk.transforms.DoFn
import org.apache.beam.sdk.values.TypeDescriptor

/**
 * Base class for the custom DoFn allowing the interaction with lifecycle methods.
 *
 * This class gives shared elements like type descriptors, input and output
 *
 * The class that extend [fr.groupbees.midgard.transforms.BaseElementFn] must give input and output type descriptors and
 * uses the tuple tags to handle good output and failure.
 *
 * @param <InputT>  input of [org.apache.beam.sdk.transforms.DoFn]
 * @param <OutputT> output of [org.apache.beam.sdk.transforms.DoFn]
 * @author mazlum
 */
abstract class BaseElementFn<InputT, OutputT> : DoFn<InputT, OutputT> {

    @JvmField
    @Transient
    protected var inputType: TypeDescriptor<InputT>

    @JvmField
    @Transient
    protected var outputType: TypeDescriptor<OutputT>

    /**
     *
     * Constructor for BaseElementFn.
     */
    protected constructor() {
        inputType = super.getInputTypeDescriptor()
        outputType = super.getOutputTypeDescriptor()
    }

    /**
     * Constructor for BaseElementFn.
     *
     * @param inputType  a [org.apache.beam.sdk.values.TypeDescriptor] object
     * @param outputType a [org.apache.beam.sdk.values.TypeDescriptor] object
     */
    protected constructor(inputType: TypeDescriptor<InputT>, outputType: TypeDescriptor<OutputT>) {
        this.inputType = inputType
        this.outputType = outputType
    }

    /**
     *
     * Getter for the field `inputType`.
     *
     * @return a [org.apache.beam.sdk.values.TypeDescriptor] object
     */
    override fun getInputTypeDescriptor(): TypeDescriptor<InputT> {
        return inputType
    }

    /**
     *
     * Getter for the field `outputType`.
     *
     * @return a [org.apache.beam.sdk.values.TypeDescriptor] object
     */
    override fun getOutputTypeDescriptor(): TypeDescriptor<OutputT> {
        return outputType
    }
}